lib/sqlalchemy/ext/automap.py

```diff
@@ -670,9 +736,29 @@ class to another, for a collection reference.
     return referred_cls.__name__.lower() + "_collection"
 
 
+class GenerateRelationshipType(Protocol):
+    def __call__(
+        self,
+        base: Type[Any],
+        direction: RelationshipDirection,
+        return_fn: Callable[..., Union[Relationship[Any], ORMBackrefArgument]],
+        attrname: str,
+        local_cls: Type[Any],
+        referred_cls: Type[Any],
+        **kw: Any,
+    ) -> Union[ORMBackrefArgument, Relationship[Any]]:
+        ...
+
+
 def generate_relationship(
-    base, direction, return_fn, attrname, local_cls, referred_cls, **kw
-):
+    base: Type[Any],
+    direction: RelationshipDirection,
+    return_fn: Callable[..., Union[Relationship[Any], ORMBackrefArgument]],
```

Comment:
**mike bayer** (zzzeek) wrote:

so, subtlety here, it's passed a callable that returns relationship, or a callable that returns ormbackref argument.   as opposed to a callable that returns a relationship or ormbackref argument.  The former will allow us to build overloads here.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4271